x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/AutoFollowCoordinator.java

```diff
@@ -172,6 +174,14 @@ public synchronized AutoFollowStats getStats() {
     }
 
     synchronized void updateStats(List<AutoFollowResult> results) {
+        // purge stats for removed patterns
+        var currentPatterns = this.patterns;
+        recentAutoFollowErrors.keySet().removeIf(key -> {
+            var semicolonPosition = key.indexOf(':');
+            var patternName = semicolonPosition != -1 ? key.substring(0, semicolonPosition) : key;
+            return currentPatterns.contains(patternName) == false;
```

Comment:
Because patterns can contain `:` (if I am right about that), we need to check if currentPatterns either contain the `patternName` or the `key` and keep it either way.

Reply:
Updated, though I wonder if it is better to change the key type to instead to avoid ambiguity?

Maybe something like:
```
class ErrorKey(String pattern, @Nullable String index)
```
though it would require transforming entire map when reading errors

Reply:
Yes, I agree to this, would be better to change they key type. I prefer the explicit key over using a tuple, since then we can verify that the `pattern` is not null, whereas the other is allowed to be null. And we can make the string concat of `pattern + ':' + index` inside the explicit key class. Feel free to either include into this PR or tackle in a follow-up - as you please.